src/ctypes-foreign-base/ctypes_ffi.ml

```diff
@@ -221,13 +221,78 @@ struct
     fun f ->
       let boxed = cs (Ctypes_weak_ref.make f) in
       let id = Closure_properties.record (Obj.repr f) (Obj.repr boxed) in
-      let funptr = Ctypes_ffi_stubs.make_function_pointer cs' id in
+      Ctypes_ffi_stubs.make_function_pointer cs' id
+
+  let pointer_of_function ~abi ~acquire_runtime_lock ~thread_registration fn =
+    let make_funptr = pointer_of_function_internal ~abi ~acquire_runtime_lock ~thread_registration fn in
+    fun f ->
+      let funptr = make_funptr f in
       (* TODO: use a more intelligent strategy for keeping function pointers
          associated with top-level functions alive (e.g. cache function
          pointer creation by (function, type), or possibly even just by
          function, since the C arity and types must be the same in each case.)
-         See the note by [kept_alive_indefinitely].  *)
+         See the note by [kept_alive_indefinitely].
+
+         [dynamic_funptr_of_fun] below allows for explicit life cycle management
+         and is probably safer to use. *)
       let () = keep_alive funptr ~while_live:f in
       funptr_of_rawptr fn
         (Ctypes_ffi_stubs.raw_address_of_function_pointer funptr)
+
+  type 'a funptr =
+    { mutable gc_root : unit Ctypes.ptr
+    ; fn : 'a Ctypes.static_funptr
+    ; debug_info : string option
```

Comment:
From experience I've found it extremely hard to locate the source of not correctly managed funptr without the debug_info, especially in large programs.

I'm not sure, but wouldn't we loose the ability to detect failure to free if the [funptr] was the gc root?

Reply:
> From experience I've found it extremely hard to locate the source of not correctly managed funptr without the debug_info, especially in large programs.

Yes, I can easily believe that.  I can imagine that a separate "debug" implementation of `Ctypes_ffi` that includes this extra information could be very handy.  (It could even be useful to have a general mechanism in the OCaml compiler for optionally attaching origin information when allocating values so that they can be inspected later on in the program.)

But including this information for every function pointer in every program introduces more overhead than I'm comfortable with.  So for the moment I think we should take it out.

> I'm not sure, but wouldn't we loose the ability to detect failure to free if the [funptr] was the gc root?

What I have in mind is keeping both the `gc_root` and the `fn` value, but storing the `gc_root` in the [`managed`](https://github.com/ocamllabs/ocaml-ctypes/blob/7a5656807d54bdeefd2a17d5709b30864547b4bf/src/ctypes/ctypes_ptr.ml#L58) slot that pointers already carry, to avoid allocating an additional record here.  That would require some fairly intrusive library-wide changes, so I don't propose doing it in this PR.

Reply:
I'm convinced, `debug_info` has been removed.